### ISSUE TITLE
Buffer overflow when executing lxc attach command

### ISSUE THREAD

**Issue by banderson5586** [lxc/lxc#2123]
I'm running many LXC containers to simulate network devices, the python application that wraps these also starts the network services (zebra, ospfd, bgpd, lldpd).  When executed from outside the context of the application the following would work:

cmd = "lxc-attach -n {0} -- zebra -d &"
Popen(cmd)

but when run in the application after starting up the LXC containers I receive a buffer overflow. What is odd is that I was using a smaller simulation definition and the issue was not preset. When I switched to a larger simulation the issue is present. 

# Required information

 * Distribution: Ubuntu and RHEL
 * The output of
   * verion = 1.1.5
   *  python verion = .1
   *  uname = 4.13.0-32-generic
   * 12:net_cls,net_prio:/
      11:cpu,cpuacct:/user.slice/user-1000.slice
      10:cpuset:/
      9:rdma:/
      8:perf_event:/
      7:hugetlb:/
      6:devices:/user.slice/user-1000.slice
      5:pids:/user.slice/user-1000.slice
      4:memory:/user/benander/0
      3:blkio:/user.slice/user-1000.slice
      2:freezer:/user/benander/0
      1:name=systemd:/user.slice/user-1000.slice/session-41.scope
   * sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,relatime,size=8139488k,nr_inodes=2034872,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=1633172k,mode=755 0 0
/dev/sda2 / ext4 rw,relatime,errors=remount-ro,data=ordered 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
tmpfs /sys/fs/cgroup tmpfs rw,mode=755 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids,release_agent=/run/cgmanager/agents/cgm-release-agent.pids 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb,release_agent=/run/cgmanager/agents/cgm-release-agent.hugetlb 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event,release_agent=/run/cgmanager/agents/cgm-release-agent.perf_event 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma,release_agent=/run/cgmanager/agents/cgm-release-agent.rdma 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset,clone_children 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=27,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=19491 0 0
mqueue /dev/mqueue mqueue rw,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,relatime 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime,pagesize=2M 0 0
fusectl /sys/fs/fuse/connections fusectl rw,relatime 0 0
configfs /sys/kernel/config configfs rw,relatime 0 0
tracefs /sys/kernel/debug/tracing tracefs rw,relatime 0 0
/dev/loop0 /snap/core/3748 squashfs ro,nodev,relatime 0 0
/dev/loop1 /snap/ubports-installer/6 squashfs ro,nodev,relatime 0 0
/dev/loop2 /snap/core/3887 squashfs ro,nodev,relatime 0 0
/dev/loop3 /snap/core/3604 squashfs ro,nodev,relatime 0 0
/dev/sda1 /boot/efi vfat rw,relatime,fmask=0077,dmask=0077,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,relatime 0 0
cgmfs /run/cgmanager/fs tmpfs rw,relatime,size=100k,mode=755 0 0
lxcfs /var/lib/lxcfs fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=1633172k,mode=700,uid=1000,gid=1000 0 0
gvfsd-fuse /run/user/1000/gvfs fuse.gvfsd-fuse rw,nosuid,nodev,relatime,user_id=1000,group_id=1000 0 0
/dev/sda2 /var/lib/docker/overlay2 ext4 rw,relatime,errors=remount-ro,data=ordered 0 0
/dev/loop4 /mnt vfat rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro 0 0
tmpfs /run/user/108 tmpfs rw,nosuid,nodev,relatime,size=1633172k,mode=700,uid=108,gid=114 0 0
gvfsd-fuse /run/user/108/gvfs fuse.gvfsd-fuse rw,nosuid,nodev,relatime,user_id=108,group_id=114 0 0
overlay /var/lib/docker/overlay2/6aa84858ecb6207d43ab9e5006127fffb211c4d4857bc254ba1d22d646a32ab8/merged overlay rw,relatime,lowerdir=l/ENAQTZZ7GMOXGTMTBIIQV3OT7Z:l/BLX4U46FANCC63ZGEAINRAWBMR:l/QIDBPBG4JJGPUYIW3MIEM55SX4:l/Z63BT262ELGROVWHSPHKRFVTBA:l/XK473ETMCSXMBKMEEEDWN6XG3I:l/ZGVUKRZX4CXSDS5LFIMUEQLLHA:l/TGKTIEBRXG35SXLPGIGZKGL2JO:l/RGRRSGTZICILRTY2FQPYKQ2F7F:l/OP2VIXQU2SJBSZCXO2BB3YWKJ7:l/OGXGL72HSLIB2LGSTXKGFX5XX2:l/TM337MLCUI57ZDFOUJ5CBAY6EM:l/3DUIABZGCTZUD7NF635EGZHF3R:l/TD3H3DCHQHDTRJCVA425GPGBQP:l/ME7FN3JGCOLD3XBL5K7ZQVRX6K:l/IKBDKRT77X72XINUZFXKFDERQT:l/II3LO5KVOTGCFBE4ZGWLNAVQQM:l/BPF7EDOVD2MWANZ2AKYPPX6KD4:l/IN2KK2OO4PLG273O3O6I2XIO7L:l/RWU2OKMPNG2A6MZPPJO3UEBTGA,upperdir=6aa84858ecb6207d43ab9e5006127fffb211c4d4857bc254ba1d22d646a32ab8/diff,workdir=6aa84858ecb6207d43ab9e5006127fffb211c4d4857bc254ba1d22d646a32ab8/work 0 0
shm /var/lib/docker/containers/d9b4386f325df657e44961e454acf95198a4d86deb685f0a5a19045e2e902739/shm tmpfs rw,nosuid,nodev,noexec,relatime,size=65536k 0 0

[overflow.txt](https://github.com/lxc/lxc/files/1688072/overflow.txt)

**Comment by brauner** [comment]
That's not `lxc-attach` segfaulting that's this `zebra` command segfaulting:

```(Pdb) *** buffer overflow detected ***: zebra terminated
======= Backtrace: =========
/lib64/libc.so.6(__fortify_fail+0x37)[0x7f3c95233eb7]
/lib64/libc.so.6(+0x10c070)[0x7f3c95232070]
/lib64/libc.so.6(+0x10de27)[0x7f3c95233e27]
/usr/lib64/quagga/libzebra.so.0(funcname_thread_add_read+0x38)[0x7f3c9797eb38]
zebra(kernel_init+0x1b3)[0x7f3c97e114c3]
zebra(main+0x30a)[0x7f3c97df504a]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f3c95147af5]
zebra(+0xe1c1)[0x7f3c97df51c1]
======= Memory map: ========
```

:)

**Comment by thebenanderson** [comment]
I did realize that :)

I know zebra (and ospfd, and bgpd) segfaulting, but those only show this behavior in the context of using the LXC library, and only when using it in this manner, at this scale.  

If I spin up 10000 containers in the python app and attempt to start these processes it will error. 

If I spin up 10000 containers in a python app and then run a separate script with the same lxc attach_run_command calls (or popen(lxc-attach) calls) it will not error.

If I manually attach (similar to separate script) it will not error.

It's a long shot but is there any way that invoking lxc-attach from python would be linked to this? I've been troubleshooting for several days and it's only in this specific context that I can reproduce. In smaller simulations with smaller configs the issues isn't present.  A simulation with 200 devices does not suffer from this error. The topology is defined by yaml, the configs are normal text configs so there aren't any unusual variables there.

**Comment by brauner** [comment]
Is it possible for you to get hold of a newer lxc and see if you can reproduce?

**Comment by thebenanderson** [comment]
Yes. I'll try in the morning and see if the issue is still present with the latest stable release.

**Comment by thebenanderson** [comment]
I simplified my setup in the initial statement (long story as to why I'm running LXC in docker). Here is the full setup. These LXC containers are running in a RHEL docker container. The docker container runs on either my devbox/destkop (Ubuntu) or a lab machine (RHEL).  On my desktop docker container with version 1.1.5   I see the LXC containers, with 2.*.* LXC I do not.  On the lab box,  RHEL with RHEL container, I do see the containers using python3 LXC lib. 

I will test the buffer overflow with the new version on my lab host but would be interested to know why I don't see the containers when run on my desktop (Ubuntu host running RHEL docker). 

Here is the LXC config, python and shell commands I used that show the container process is running but LXC commands don't recognize it. I also tried with LXC 2.1.1. If you believe this is a different bug let me know and I'll provide more info.

lxc.init_cmd=/bin/pause
lxc.utsname=dev1
lxc.network.type=empty
lxc.rootfs=overlayfs:/lxcroot/testfs:/lxcroot/dev1/local
lxc.mount.auto=proc sys
lxc.pts=1
lxc.autodev=1
lxc.kmsg=0


Python 3.4.5 (default, Dec 11 2017, 14:22:24) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-16)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import lxc
>>> con = lxc.Container('dev1')
>>> con.set_config_path('/lxcroot')
True
>>> con.load_config()
False
>>> con.load_config()
False
>>> con = lxc.Container('dev1')
>>> con.set_config_path('/lxcroot')
True
>>> con.load_config()
True
>>> con.start(daemonize=True, close_fds=True)
True
>>> con.init_pid
153


[root@f68884309295 lxcroot]# lxc-start --version
2.0.6
[root@f68884309295 lxcroot]# lxc-ls --active
[root@f68884309295 lxcroot]# lxc-attach -n dev1
Error: container dev1 is not defined
[root@f68884309295 lxcroot]# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
...
root       153  0.0  0.0   4160   656 ?        S    23:33   0:00 /bin/pause
